### PR TITLE
Add Kotlin support

### LINKED ISSUE #295: Kotlin support
[Kotlin](http://kotlinlang.org/) is a programming language by JetBrains that was built as a better Java.

### PR DESCRIPTION
**Pull request by meleyal**
Fixes #295
Extends #354

### CODE REVIEW

File: lib/rouge/lexers/kotlin.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        object open operator out override package private protected
+        public reified return sealed set super tailrec this throw
+        true try typealias typeof val var vararg when where while
+        yield
+      )
+
+      name = '@?[_\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Nl}]' +
+             '[\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Nl}\p{Nd}\p{Pc}\p{Cf}\p{Mn}\p{Mc}]*'
+
+      id = "(#{name}|`#{name}`)"
```
Comment (jneen) [inline lib/rouge/lexers/kotlin.rb]: I'd prefer if `name` and `id` here were themselves regexes - Ruby has some safety features when interpolating regexes within other ones that I'd like to take advantage of.

**Comment by jneen** [comment]
Otherwise, looks really good, thanks for the lexer!

File: lib/rouge/lexers/kotlin.rb (hunk 1 above)
Comment (meleyal) [inline lib/rouge/lexers/kotlin.rb]: @jneen, no problem, but could you clarify your point about the safety features?

Comment (jneen) [inline lib/rouge/lexers/kotlin.rb]:
``` irb
=> /foo/i
irb(main):004:0> outer = /bar#{inner}baz/
=> /bar(?i-mx:foo)baz/
irb(main):005:0> 
```

More or less it automatically surrounds with non-capturing parens and preserves flags. Backslashing is also a little more straightforward.

**Comment by pabl0rg** [comment]
Hi, do you know when rouge might get Kotlin support?  Thank you

**Comment by jneen** [comment]
@pabl0rg as soon as the review process is finished, I can merge this. Currently waiting on a few changes from @meleyal.

**Comment by meleyal** [comment]
@jneen changed to use a regex for name and id, also squashed the commits - let me know if it looks ok now.

**Pull request merged**